app/wasm.go

```diff
@@ -108,3 +118,114 @@ type AllPriceResponse struct {
+	Routes  []ammtype.SwapAmountInRoute `protobuf:"bytes,1,rep,name=routes,proto3" json:"routes,omitempty"`
+}
+
+type QuerySwapEstimationResponse struct {
+	SpotPrice sdk.Dec  `protobuf:"bytes,1,opt,name=SpotPrice,proto3" json:"spot_price,omitempty"`
+	TokenOut  sdk.Coin `protobuf:"bytes,2,opt,name=tokenOut,proto3" json:"token_out,omitempty"`
+}
+
+func CustomMessageDecorator(amm *ammkeeper.Keeper) func(wasmkeeper.Messenger) wasmkeeper.Messenger {
+	return func(old wasmkeeper.Messenger) wasmkeeper.Messenger {
+		return &CustomMessenger{
+			wrapped: old,
+			amm:     amm,
+		}
+	}
+}
+
+type CustomMessenger struct {
+	wrapped wasmkeeper.Messenger
+	amm     *ammkeeper.Keeper
+}
+
+var _ wasmkeeper.Messenger = (*CustomMessenger)(nil)
+
+func (m *CustomMessenger) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddress, contractIBCPortID string, msg wasmvmtypes.CosmosMsg) ([]sdk.Event, [][]byte, error) {
+	if msg.Custom != nil {
+		// only handle the happy path where this is really creating / minting / swapping ...
+		// leave everything else for the wrapped version
+		var contractMsg ElysMsg
+		if err := json.Unmarshal(msg.Custom, &contractMsg); err != nil {
+			return nil, nil, errorsmod.Wrap(err, "elys msg")
+		}
+		if contractMsg.MsgSwapExactAmountIn != nil {
+			return m.msgSwapExactAmountIn(ctx, contractAddr, contractMsg.MsgSwapExactAmountIn)
+		}
+	}
+	return m.wrapped.DispatchMsg(ctx, contractAddr, contractIBCPortID, msg)
+}
+
+func (m *CustomMessenger) msgSwapExactAmountIn(ctx sdk.Context, contractAddr sdk.AccAddress, msgSwapExactAmountIn *MsgSwapExactAmountIn) ([]sdk.Event, [][]byte, error) {
+	res, err := PerformMsgSwapExactAmountIn(m.amm, ctx, contractAddr, msgSwapExactAmountIn)
+	if err != nil {
+		return nil, nil, errorsmod.Wrap(err, "perform swap")
+	}
+
+	responseBytes, err := json.Marshal(*res)
+	if err != nil {
+		return nil, nil, errorsmod.Wrap(err, "failed to serialize swap response")
+	}
+
+	resp := [][]byte{responseBytes}
+
+	return nil, resp, nil
+}
+
+func PerformMsgSwapExactAmountIn(f *ammkeeper.Keeper, ctx sdk.Context, contractAddr sdk.AccAddress, msgSwapExactAmountIn *MsgSwapExactAmountIn) (*MsgSwapExactAmountInResponse, error) {
+	if msgSwapExactAmountIn == nil {
+		return nil, wasmvmtypes.InvalidRequest{Err: "swap null swap"}
+	}
+
+	msgServer := ammkeeper.NewMsgServerImpl(*f)
+
+	var PoolIds []uint64
+	var TokenOutDenoms []string
+
+	for _, route := range msgSwapExactAmountIn.Routes {
+		PoolIds = append(PoolIds, route.PoolId)
+		TokenOutDenoms = append(TokenOutDenoms, route.TokenOutDenom)
+	}
+
+	msgMsgSwapExactAmountIn := ammtype.NewMsgSwapExactAmountIn(msgSwapExactAmountIn.Sender, msgSwapExactAmountIn.TokenIn, msgSwapExactAmountIn.TokenOutMinAmount, PoolIds, TokenOutDenoms)
+
+	if err := msgMsgSwapExactAmountIn.ValidateBasic(); err != nil {
+		return nil, errorsmod.Wrap(err, "failed validating MsgMsgSwapExactAmountIn")
+	}
+
+	// Swap
+	swap_resp, err := msgServer.SwapExactAmountIn(
```

Comment:
Can you update the variable name to `swap_resp` to `swapResp`? We use camelcase on variable names for chain side codebase